benchrun/python/benchrun/_benchmark.py

```diff
@@ -0,0 +1,280 @@
+        element, will override any keys that already exist, i.e. it does not append
+        recursively.
+
+        In most cases, should be set on the adapter, not here, but can be set here to
+        specify benchmark-dependent values like appending a tag for ``suite``.
+    drop_caches : bool
+        Try to drop disk caches?
+    gc_collect : bool
+        Run garbage collection before timing code?
+    gc_disable : bool
+        Disable garbage collection during timing?
+    error_handling : str
+        What should happen if an iteration errors out? Options: ``"stop"`` (skip future iterations
+        and report only error), ``"break"`` (skip future iterations and report everything so far),
+        ``"continue"`` (run all iterations even if they fail and report everything). As we currently
+        can't report both metrics and errors for the same benchmark, ``"stop"`` and ``"break"`` are
+        currently identical, and ``"continue"`` may run longer, but will report the same thing.
+    """
+
+    iteration: Iteration = None
+    case_list: CaseList = None
+    result_fields_append: Dict[str, Any] = None
+    settings = None
+    cache = None
+
+    def __init__(
+        self,
+        iteration: Iteration,
+        case_list: CaseList,
+        result_fields_append: Dict[str, Any] = {},
+        drop_caches: bool = False,
+        gc_collect: bool = True,
+        gc_disable: bool = True,
+        error_handling: str = "stop",
+    ) -> None:
+        assert error_handling in ["stop", "break", "continue"]
+
+        self.iteration = iteration
+        self.case_list = case_list
+        self.result_fields_append = result_fields_append or {}
+        self.settings = {
+            # TODO: Figure out if we should change `drop_caches` default to True
+            "drop_caches": drop_caches,
+            "gc_collect": gc_collect,
+            "gc_disable": gc_disable,
+            "error_handling": error_handling,
+        }
+        self.cache = CacheManager()
+
+    def run_case(
+        self,
+        case: dict,
+        iterations: int,
+        run_reason: str,
+        run_name: str,
+        run_id: str,
+        batch_id: str,
+    ) -> BenchmarkResult:
+        iteration = copy.deepcopy(self.iteration)
+        queue = mp.Queue()
+        proc = mp.Process(
+            target=iteration, args=(case, iterations, self.settings, queue)
+        )
+        proc.start()
+        res = queue.get()
+        proc.join()
+
+        result = BenchmarkResult(
+            run_name=run_name,
+            run_id=run_id,
+            batch_id=batch_id,
+            run_reason=run_reason,
+            stats=res["stats"],
+            error=res["error"],
+            tags={"name": self.iteration.name, **case},
+            # info={},  # TODO: is there common detectable metadata worth putting here?
+            context={"benchmark_language": "Python"},
+        )
+
+        for param in self.result_fields_append:
```

Comment:
Isn't this code already in the BenchmarkAdapter? Why do we have to duplicate it here?

Reply:
(probably that CI README example will help me figure this out)

Reply:
Yep, and I don't really love the duplication, but the reason is that there will be one `CallableAdapter` instance for all the benchmarks in the `BenchmarkList`, but you may want to set things on some benchmarks but not others, e.g. say a tag for something about the data format for a particular benchmark that's the same for everything now (so it's not necessarily parameterized and therefore a tag) but which may vary in the future.

I could separate it out into a `update_benchmark_result()` util function in `benchadapt` so the same code gets used in both places? Though this isn't huge or likely to change

Reply:
yeah maybe we can leave it for now until the packages are more stable